build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/MrjarPlugin.java

```diff
@@ -60,46 +60,33 @@ public void apply(Project project) {
-        try (var subdirStream = Files.list(srcDir)) {
-            for (Path sourceset : subdirStream.toList()) {
-                assert Files.isDirectory(sourceset);
-                String sourcesetName = sourceset.getFileName().toString();
-                Matcher sourcesetMatcher = MRJAR_SOURCESET_PATTERN.matcher(sourcesetName);
-                if (sourcesetMatcher.matches()) {
-                    mainVersions.add(Integer.parseInt(sourcesetMatcher.group(1)));
-                }
-            }
-        } catch (IOException e) {
-            throw new UncheckedIOException(e);
-        }
-
-        Collections.sort(mainVersions);
-        List<String> parentSourceSets = new ArrayList<>();
-        parentSourceSets.add(SourceSet.MAIN_SOURCE_SET_NAME);
+        List<Integer> mainVersions = findSourceVersions(project);
+        List<String> mainSourceSets = new ArrayList<>();
+        mainSourceSets.add(SourceSet.MAIN_SOURCE_SET_NAME);
+        List<String> testSourceSets = new ArrayList<>(mainSourceSets);
+        testSourceSets.add(SourceSet.TEST_SOURCE_SET_NAME);
         for (int javaVersion : mainVersions) {
-            String sourcesetName = "main" + javaVersion;
-            addMrjarSourceset(project, javaExtension, sourcesetName, parentSourceSets, javaVersion);
-            parentSourceSets.add(sourcesetName);
+            String mainSourceSetName = SourceSet.MAIN_SOURCE_SET_NAME + javaVersion;
+            addSourceSet(project, javaExtension, mainSourceSetName, mainSourceSets, javaVersion);
+            mainSourceSets.add(mainSourceSetName);
+            testSourceSets.add(SourceSet.MAIN_SOURCE_SET_NAME + javaVersion);
+
+            String testSourceSetName = SourceSet.TEST_SOURCE_SET_NAME + javaVersion;
+            SourceSet testSourceSet = addSourceSet(project, javaExtension, testSourceSetName, testSourceSets, javaVersion);
+            testSourceSets.add(testSourceSetName);
+            createTestTask(project, testSourceSet, javaVersion);
         }
-    }
 
-    private void addMrjarSourceset(
-        Project project,
-        JavaPluginExtension javaExtension,
-        String sourcesetName,
-        List<String> parentSourceSets,
-        int javaVersion
-    ) {
-        SourceSet sourceSet = javaExtension.getSourceSets().maybeCreate(sourcesetName);
-        for (String parentSourceSetName : parentSourceSets) {
-            GradleUtils.extendSourceSet(project, parentSourceSetName, sourcesetName);
-        }
+        configureMrjar(project, javaExtension, mainVersions, "main");
+    }
 
+    private void configureMrjar(Project project, JavaPluginExtension javaExtension, List<Integer> versions, String baseSourceSetName) {
         var jarTask = project.getTasks().withType(Jar.class).named(JavaPlugin.JAR_TASK_NAME);
         jarTask.configure(task -> {
-            task.into("META-INF/versions/" + javaVersion, copySpec -> copySpec.from(sourceSet.getOutput()));
+            for (var version : versions) {
```

Comment:
Looking at this more, this feels a bit awkward. I feel like we should be doing this in `addSourceSet` instead.